File: packages/api/src/videos.ts
import {client} from "./client";

const api = client.extend((options) => ({
    prefix: `${options.prefix}/videos`,
}));

export interface VideoResponse {
    id: number;
    label: string;
    filename: string;
    description: string;
}

export interface UpdateVideoRequest {
    label?: string | null;
    description?: string | null;
}

export const videosApi = {
    uploadVideo: async (
        label: string,
        description: string,
        file: File,
    ): Promise<VideoResponse> => {
        const formData = new FormData();
        formData.append("label", label);
        formData.append("description", description);
        formData.append("file", file);

        return await api.post("upload", {body: formData}).json();
    },

    listVideos: async (): Promise<VideoResponse[]> => {
        return await api.get("").json();
    },

    getVideo: async (videoId: number): Promise<VideoResponse> => {
        return await api.get(`${videoId}`).json();
    },

    updateVideo: async (
        videoId: number,
        request: UpdateVideoRequest,
    ): Promise<VideoResponse> => {
        return await api.patch(`${videoId}`, {json: request}).json();
    },

    deleteVideo: async (videoId: number): Promise<void> => {
        await api.delete(`${videoId}`);
    },
};

File: apps/backend/src/api/routes/videos.py
from fastapi import APIRouter, File
from fastapi import UploadFile
from fastapi.params import Depends, Form
from starlette import status

from api.dependencies import get_video_service
from models.videos import VideoResponse, UpdateVideoRequest, CreateVideoRequest
from services.video_service import VideoService

VIDEO_PREFIX = "/videos"

router = APIRouter(prefix=VIDEO_PREFIX, tags=["videos"])


@router.post("/upload", response_model=VideoResponse)
def upload_video(
        label: str = Form(...),
        description: str = Form(...),
        project_id: int = Form(...),
        file: UploadFile = File(...),
        video_service: VideoService = Depends(get_video_service)
):
    return video_service.upload_video(
        CreateVideoRequest(label=label, description=description, project_id=project_id),
        file
    )


@router.get("", response_model=list[VideoResponse])
def list_videos(video_service: VideoService = Depends(get_video_service)):
    return video_service.list_videos()


@router.get("/{video_id}", response_model=VideoResponse)
def get_video(video_id: int, video_service: VideoService = Depends(get_video_service)):
    return video_service.get_video(video_id)


@router.patch("/{video_id}", response_model=VideoResponse | None)
def update_video(
        video_id: int,
        request: UpdateVideoRequest,
        video_service: VideoService = Depends(get_video_service)
):
    return video_service.update_video(video_id, request)


@router.delete("/{video_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_video(video_id: int, video_service: VideoService = Depends(get_video_service)):
    video_service.delete_video(video_id)
